pkg/sdk/functions_ext.go

```diff
@@ -69,9 +92,135 @@ func functionDetailsFromRows(rows []FunctionDetail) (*FunctionDetails, error) {
+		p, err := parseStageLocationPath(imp)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse imports from Snowflake: %s, err: %w", *details.Imports, err)
+		}
+		functionDetailsImports = append(functionDetailsImports, *p)
+	}
+	return functionDetailsImports, nil
+}
+
+func parseStageLocationPath(location string) (*NormalizedPath, error) {
+	log.Printf("[DEBUG] parsing stage location path part: %s", location)
+	idx := strings.Index(location, "/")
+	if idx < 0 {
+		return nil, fmt.Errorf("part %s cannot be split into stage and path", location)
+	}
+	stageRaw := strings.TrimPrefix(strings.TrimSpace(location[:idx]), "@")
+	if stageRaw != "~" {
+		stageId, err := ParseSchemaObjectIdentifier(stageRaw)
+		if err != nil {
+			return nil, fmt.Errorf("part %s contains incorrect stage location: %w", location, err)
+		}
+		stageRaw = stageId.FullyQualifiedName()
+	}
+	pathRaw := strings.TrimPrefix(strings.TrimSpace(location[idx:]), "/")
+	if pathRaw == "" {
+		return nil, fmt.Errorf("part %s contains empty path", location)
+	}
+	return &NormalizedPath{stageRaw, pathRaw}, nil
+}
+
+func parseFunctionOrProcedureReturns(returns string) (datatypes.DataType, bool, error) {
+	var returnNotNull bool
+	trimmed := strings.TrimSpace(returns)
+	if strings.HasSuffix(trimmed, " NOT NULL") {
+		returnNotNull = true
+		trimmed = strings.TrimSuffix(trimmed, " NOT NULL")
+	}
+	dt, err := datatypes.ParseDataType(trimmed)
+	return dt, returnNotNull, err
+}
+
+// Format in Snowflake DB is: (argName argType, argName argType, ...).
+func parseFunctionOrProcedureSignature(signature string) ([]NormalizedArgument, error) {
+	normalizedArguments := make([]NormalizedArgument, 0)
+	trimmed := strings.TrimSpace(signature)
+	if trimmed == "" {
+		return normalizedArguments, fmt.Errorf("could not parse signature from Snowflake: %s, can't be empty", signature)
+	}
+	if trimmed == "()" {
+		return normalizedArguments, nil
+	}
+	if !strings.HasPrefix(trimmed, "(") || !strings.HasSuffix(trimmed, ")") {
+		return normalizedArguments, fmt.Errorf("could not parse signature from Snowflake: %s, wrapping parentheses not found", trimmed)
+	}
+	raw := (trimmed)[1 : len(trimmed)-1]
+	args := strings.Split(raw, ",")
+
+	for _, arg := range args {
+		a, err := parseFunctionOrProcedureArgument(arg)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse signature from Snowflake: %s, err: %w", trimmed, err)
+		}
+		normalizedArguments = append(normalizedArguments, *a)
+	}
+	return normalizedArguments, nil
+}
+
+// TODO [SNOW-1850370]: test with strange arg names (first integration test)
+func parseFunctionOrProcedureArgument(arg string) (*NormalizedArgument, error) {
+	log.Printf("[DEBUG] parsing argument: %s", arg)
+	trimmed := strings.TrimSpace(arg)
+	idx := strings.Index(trimmed, " ")
+	if idx < 0 {
+		return nil, fmt.Errorf("arg %s cannot be split into arg name, data type, and default", arg)
+	}
+	argName := trimmed[:idx]
+	rest := strings.TrimSpace(trimmed[idx:])
```

Comment:
🤔  isn't `DEFAULT` as part of signature (after data type)?

Reply:
It's not, check `create function for Java - default argument value` - it's part of the SHOW (arguments) but not a part of DESCRIBE's signature. Still: it does not provide a value, so the only thing we can do in the future is to use this DEFAULT from SHOW to discover external changes in adding/removing the DEFAULT overall.

Reply:
Kk I was just wondering what we are parsing here, if the input is coming from describe's signature then it's fine. I was just wondering if `strings.TrimSpace(trimmed[idx:])` could possibly encounter something more than just data type.